### ISSUE TITLE
Container name escaping doesn't align with latest Docker rules

### ISSUE THREAD

**Issue by jfly** [jupyterhub/dockerspawner#326]
(Possibly related: https://github.com/jupyterhub/dockerspawner/issues/205, https://github.com/jupyterhub/dockerspawner/issues/305)

## Some history + the problem

1. In https://github.com/jupyterhub/dockerspawner/pull/40, `dockerspawner` gained the ability to escape "special characters" in container names.
2. ^ This commit was built in reaction to this change to Docker to require that container names match the regex `^/?[a-zA-Z0-9_-]+$`: https://github.com/moby/moby/commit/f63cdf0260cf6287d28a589a79d3f947def6a569

3. Docker has since changed their container naming rules a couple times:
   - Allowing periods: https://github.com/moby/moby/commit/1940015824f
   - Requiring all container names to start with `a-zA-Z0-9`: https://github.com/moby/moby/commit/aa2af817bee123827f4a857555db8fea5e01e7d7

Currently, Docker requires container names to match `^[a-zA-Z0-9][a-zA-Z0-9_.-]$`: https://github.com/moby/moby/blob/928381b2215c9608692e663e6ddfa9e609bc2d6f/daemon/names/names.go#L5-L6, which is is different from [`dockerspawner`'s current escaping logic](https://github.com/jupyterhub/dockerspawner/blob/9d4a35995d2c2dd992e070cc7ad260123308b606/dockerspawner/dockerspawner.py#L672-L678) in two important ways:
  1. `dockerspawner` unnecessarily treats `.` as unsafe. You could say this isn't exactly a bug, but it is a bit weird.
  2. `_`, `.`, and `-` are only acceptable container names if they are *not* the first character in the container name. Right now, if someone's username starts with a special character, `dockerspawner` will escape that special character and generate an escaped container name that starts with a `_`. For example:
     ```
     $ pip freeze | grep -i dockerspawner
     dockerspawner==0.11.1
 
     $ python
     Python 3.7.4 (default, Jul 16 2019, 07:12:58) 
     >>> import dockerspawner
     >>> ds = dockerspawner.DockerSpawner()
     >>> ds._escape("$Money Name$")
     '_24Money_20Name_24'
     ```
     Proof that this container name won't work with docker:
 
     ```
     $ docker run --name "_24Money_20Name_24" ubuntu
     docker: Error response from daemon: Invalid container name (_24Money_20Name_24), only [a-zA-Z0-9][a-zA-Z0-9_.-] are allowed.
     See 'docker run --help'.
     ```

## Proposals

I think the right thing to do would be for `dockerspawnner` to add `.` to the `_docker_safe_chars` list (to match Docker's current behavior), and also update the escaping logic to never generate `_` as the first character in a name. Some ideas:

1. Change [`_docker_escape_char`](https://github.com/jupyterhub/dockerspawner/blob/9d4a35995d2c2dd992e070cc7ad260123308b606/dockerspawner/dockerspawner.py#L525) from `_` to something that *is* an allowed first character container name.
   - This is pretty simple, but all the choices we could make (`[a-zA-Z0-9]`) are pretty unfortunate.
2. Prefix all container names with something known to be safe. Something like "dockerspawner-"?
   - This is pretty simple, and probably less weird than 1) above.
3. Handle the first character of the container name specially in some weird way. Maybe say "we'll use `0` as an escape character for the *first* character of the name, and then use `_` as the escape character for the rest of the name.
   - Cons: This is pretty weird. The decoding logic would also be weird, but I'm not sure that `dockerspawner` actually ever decodes the container names it produces.
   - Pros: It's the most similar of all these options to the current behavior, but still obeys all of Docker's current container naming rules.

Happy to send in a PR if someone wants to give guidance on which of 1, 2, 3 above sound good.

**Comment by matteoipri** [comment]
Hi, I actually encountered the issue where SwarmSpawner was not able to spawn a docker service/container for the user because the container name was not a valid domain name.
Thus, actually _ can not be used and should be escaped itself.
I found [this](http://www.netregister.biz/faqit.htm) where at the first question it is stated that a domain name must match the regular expression [A-Za-z0-9-]{1,63}.

This is what I did to overcome the issue:
```python
class MISwarmSpawner(SwarmSpawner):
    """docstring for MISwarmSpawner"""

    def template_namespace(self):
        escaped_image = self.image.replace("/", "_")
        server_name = getattr(self, "name", "")
        safe_username = self.escaped_name.replace("-", "--").replace("_", "-").lower()
        return {
            "username": self.escaped_name,
            "safe_username": safe_username,
            "raw_username": self.user.name,
            "imagename": escaped_image,
            "servername": server_name,
            "prefix": self.prefix,
        }
```
Basically I am proposing to use - instead of _ as an escape character.
In my implementation the character - itself is escaped doubling it (as you would do in C for the % character) and the other _percent_ escaped characters use - instead of _ (or %).

Since I have a prefix set in the hub configuration file, and the prefix is made of only lowercase letters, there's no need to treat the first character of the username in a special way, when using both DockerSpawner and SwarmSpawner.

To answer @jfly 's proposals, I like 2, which is what is already done, but it is not enough to fix all issues, IMHO.
